File: examples/rust/examples/simple_crud.rs
use reqwest::header::{HeaderMap, HeaderValue, AUTHORIZATION, CONTENT_TYPE};
use serde_json::{json, Value};
use std::env;

const COLLECTION: &str = "simple_crud_example_rs";

async fn delete_collection(
    client: &reqwest::Client,
    base_url: &str,
) -> Result<(), Box<dyn std::error::Error>> {
    let response = client
        .delete(format!("{base_url}/api/collections/{COLLECTION}"))
        .send()
        .await?;
    if response.status() != reqwest::StatusCode::NOT_FOUND {
        response.error_for_status()?;
    }
    Ok(())
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    dotenv::dotenv().ok();

    let base_url = env::var("API_BASE_URL").unwrap_or_else(|_| "http://localhost:8080".to_string());
    let api_key =
        env::var("API_BASE_KEY").unwrap_or_else(|_| "a-test-api-key-from-ekodb".to_string());

    // Get authentication token
    let client = reqwest::Client::new();
    let token_response = client
        .post(&format!("{}/api/auth/token", base_url))
        .json(&json!({ "api_key": api_key }))
        .send()
        .await?;

    let token_data: Value = token_response.json().await?;
    let token = token_data["token"].as_str().unwrap();
    println!("✓ Authentication successful");

    // Create HTTP client with default headers
    let mut headers = HeaderMap::new();
    headers.insert(
        AUTHORIZATION,
        HeaderValue::from_str(&format!("Bearer {}", token))?,
    );
    headers.insert(CONTENT_TYPE, HeaderValue::from_static("application/json"));
    let client = reqwest::Client::builder()
        .default_headers(headers)
        .build()?;

    delete_collection(&client, &base_url).await?;
    let operation_result = run_examples(&client, &base_url).await;
    let cleanup_result = delete_collection(&client, &base_url).await;
    match (operation_result, cleanup_result) {
        (Err(primary), Err(cleanup)) => {
            return Err(format!("{primary}; cleanup also failed: {cleanup}").into());
        }
        (Err(primary), Ok(())) => return Err(primary),
        (Ok(()), Err(cleanup)) => return Err(cleanup),
        (Ok(()), Ok(())) => {}
    }

    println!("\n✓ All CRUD operations completed successfully");
    Ok(())
}

async fn run_examples(
    client: &reqwest::Client,
    base_url: &str,
) -> Result<(), Box<dyn std::error::Error>> {
    // Example 1: Insert a document
    println!("\n=== Insert Document ===");
    let doc: Value = client
        .post(format!("{base_url}/api/insert/{COLLECTION}"))
        .json(&json!({
            "name": "Test Record",
            "value": 42,
            "active": true
        }))
        .send()
        .await?
        .json()
        .await?;
    println!("Inserted: {:?}", doc);
    let doc_id = doc["id"].as_str().unwrap();

    // Example 2: Find by ID
    println!("\n=== Find by ID ===");
    let found_doc: Value = client
        .get(format!("{base_url}/api/find/{COLLECTION}/{doc_id}"))
        .send()
        .await?
        .json()
        .await?;
    println!("Found: {:?}", found_doc);

    // Example 3: Find with query
    println!("\n=== Find with Query ===");
    let docs: Value = client
        .post(format!("{base_url}/api/find/{COLLECTION}"))
        .json(&json!({
            "filter": {
                "type": "Condition",
                "content": {
                    "field": "active",
                    "operator": "Eq",
                    "value": true
                }
            },
            "limit": 10
        }))
        .send()
        .await?
        .json()
        .await?;
    println!("Found documents: {:?}", docs);

    // Example 4: Update document
    println!("\n=== Update Document ===");
    let updated: Value = client
        .put(format!("{base_url}/api/update/{COLLECTION}/{doc_id}"))
        .json(&json!({
            "name": "Updated Record",
            "value": 100
        }))
        .send()
        .await?
        .json()
        .await?;
    println!("Updated: {:?}", updated);

    // Example 5: Delete document
    println!("\n=== Delete Document ===");
    client
        .delete(format!("{base_url}/api/delete/{COLLECTION}/{doc_id}"))
        .send()
        .await?;
    println!("Deleted document");

    Ok(())
}

File: examples/go/simple_crud.go
package main

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"os"

	"github.com/joho/godotenv"
)

const simpleCRUDCollection = "simple_crud_go"

var (
	baseURL   string
	apiKey    string
	authToken string
)

func init() {
	godotenv.Load()
	baseURL = os.Getenv("API_BASE_URL")
	if baseURL == "" {
		baseURL = "http://localhost:8080"
	}
	apiKey = os.Getenv("API_BASE_KEY")
	if apiKey == "" {
		apiKey = "a-test-api-key-from-ekodb"
	}
}

func getAuthToken() (string, error) {
	if authToken != "" {
		return authToken, nil
	}

	body, _ := json.Marshal(map[string]string{"api_key": apiKey})
	resp, err := http.Post(baseURL+"/api/auth/token", "application/json", bytes.NewBuffer(body))
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	var result map[string]interface{}
	json.NewDecoder(resp.Body).Decode(&result)
	authToken = result["token"].(string)
	return authToken, nil
}

func request(method, path string, body interface{}) (map[string]interface{}, error) {
	token, err := getAuthToken()
	if err != nil {
		return nil, err
	}

	var reqBody io.Reader
	if body != nil {
		jsonBody, _ := json.Marshal(body)
		reqBody = bytes.NewBuffer(jsonBody)
	}

	req, _ := http.NewRequest(method, baseURL+path, reqBody)
	req.Header.Set("Authorization", "Bearer "+token)
	req.Header.Set("Content-Type", "application/json")

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode == 404 {
		return nil, nil
	}

	respBody, _ := io.ReadAll(resp.Body)
	if resp.StatusCode >= 300 {
		return nil, fmt.Errorf("HTTP %d: %s", resp.StatusCode, respBody)
	}
	if len(respBody) == 0 {
		return map[string]interface{}{}, nil
	}

	var result map[string]interface{}
	json.Unmarshal(respBody, &result)
	return result, nil
}

func requestArray(method, path string, body interface{}) ([]map[string]interface{}, error) {
	token, err := getAuthToken()
	if err != nil {
		return nil, err
	}

	var reqBody io.Reader
	if body != nil {
		jsonBody, _ := json.Marshal(body)
		reqBody = bytes.NewBuffer(jsonBody)
	}

	req, _ := http.NewRequest(method, baseURL+path, reqBody)
	req.Header.Set("Authorization", "Bearer "+token)
	req.Header.Set("Content-Type", "application/json")

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode == 404 {
		return nil, nil
	}

	respBody, _ := io.ReadAll(resp.Body)
	if resp.StatusCode >= 300 {
		return nil, fmt.Errorf("HTTP %d: %s", resp.StatusCode, respBody)
	}
	if len(respBody) == 0 {
		return []map[string]interface{}{}, nil
	}

	var result []map[string]interface{}
	json.Unmarshal(respBody, &result)
	return result, nil
}

func run() (runErr error) {
	fmt.Print("=== Simple CRUD Operations (Direct HTTP) ===\n\n")

	_, err := getAuthToken()
	if err != nil {
		return fmt.Errorf("auth failed: %w", err)
	}
	fmt.Println("✓ Authentication successful")
	defer func() {
		if _, err := request("DELETE", "/api/collections/"+simpleCRUDCollection, nil); err != nil {
			runErr = errors.Join(runErr, fmt.Errorf("cleanup collection %s: %w", simpleCRUDCollection, err))
		}
	}()

	// Example 1: Insert a document
	fmt.Println("\n=== Insert Document ===")
	doc, err := request("POST", "/api/insert/"+simpleCRUDCollection, map[string]interface{}{
		"name":   "Test Record",
		"value":  42,
		"active": true,
	})
	if err != nil {
		return fmt.Errorf("insert document: %w", err)
	}
	fmt.Printf("Inserted: %v\n", doc)
	docID := doc["id"].(string)

	// Example 2: Find by ID
	fmt.Println("\n=== Find by ID ===")
	foundDoc, err := request("GET", "/api/find/"+simpleCRUDCollection+"/"+docID, nil)
	if err != nil {
		return fmt.Errorf("find document: %w", err)
	}
	fmt.Printf("Found: %v\n", foundDoc)

	// Example 3: Find with query
	fmt.Println("\n=== Find with Query ===")
	docs, err := requestArray("POST", "/api/find/"+simpleCRUDCollection, map[string]interface{}{
		"filter": map[string]interface{}{
			"type": "Condition",
			"content": map[string]interface{}{
				"field":    "active",
				"operator": "Eq",
				"value":    true,
			},
		},
		"limit": 10,
	})
	if err != nil {
		return fmt.Errorf("query documents: %w", err)
	}
	fmt.Printf("Found documents: %v\n", docs)

	// Example 4: Update document
	fmt.Println("\n=== Update Document ===")
	updated, err := request("PUT", "/api/update/"+simpleCRUDCollection+"/"+docID, map[string]interface{}{
		"name":  "Updated Record",
		"value": 100,
	})
	if err != nil {
		return fmt.Errorf("update document: %w", err)
	}
	fmt.Printf("Updated: %v\n", updated)

	// Example 5: Delete document
	fmt.Println("\n=== Delete Document ===")
	if _, err := request("DELETE", "/api/delete/"+simpleCRUDCollection+"/"+docID, nil); err != nil {
		return fmt.Errorf("delete document: %w", err)
	}
	fmt.Println("Deleted document")

	fmt.Println("\n✓ All CRUD operations completed successfully")
	return nil
}

func main() {
	if err := run(); err != nil {
		fmt.Printf("✗ Simple CRUD failed: %v\n", err)
		os.Exit(1)
	}
}
